docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+This means that it cannot name any type or field.
+
+Implementation
+--------------
+
+Package-private interface and class types are simply not nameable and therefore not usable in other packages.
+The same is true if they are implicitly hidden by belonging to a version that is not in use.
+This is purely a compile-time check.
+
+Methods and fields are slightly more complex.
+Methods or fields accessed on concrete types are trivially allowed or disallowed depending on whether the caller can see an API version that includes them.
+Method or field accesses via interfaces depend on whether the selector used in the calling package unifies with the versioned selector in the callee's package.
+
+Package-private members effectively introduce an anonymous (or, at least, unnameable) namespace for selectors such that `_foo` in one package and `_foo` in another are different selectors and there is no way of code in one package naming the other selector.
+Versions simply generalise this slightly.
+Versioned members are identified by a selector that includes the version and package of origin, which has the same value as a selector of the same name and arrity in a package that imports them and is permitted to see that version.
+
+
+Corner cases / Open questions
+-----------------------------
+
+There are some corner cases that may have interesting implications here, in particular when objects are passed between modules.
+For example, consider a package `A` that exposes the following class (straw-man syntax for declaring versions, probably not what we want to end up with):
+
+```verona
+class A
+{
+	// This is always public
+	isSomething() : Bool;
+	// This is public in any API version >= 2.0
+	[[version=2.0+]]
+	isSomethingElse() : Bool;
+}
+```
+
+If a file in another package imports this then it may do something like:
+
+```verona
+import "A@1.0" as A;
+
+interface IsSomethingElse
+{
+  isSomethingElse() : Bool;
+};
+
+var a = A.A();
+```
+
+The variable `a` here does not implement the interface `IsSomethingElse` because the `isSomethingElse` method is not visible here.
+If `a` is passed to another package containing a similar interface, should it match?
+There are two obvious answers:
+
+ - If the other package imports `A` with a version <2.0, then it should not see the method.
+ - If the other package imports `A` with a version >=2.0, then it should see the method.
+
+The third case is the most interesting.
+What is the correct behaviour if the other package does not import `A` at all?
+Note that, without full dataflow analysis, it is not possible to tell in the general case which package the object was allocated in (for example, it might be stored in an array of `Any` allocated by a fourth module and populated by many other things in the system) and so the question reduces to selector identity.
+Specifically: Is the selector `isSomethingElse/1` in a module that does not import `A` the same as the selector `isSomethingElse@2.0/1` in `A`?
```

Comment:
Could we create symbol aliases for each method, field and type that points to the most concrete implementation and so each upgrade only needs to add its new aliases, because all previous ones (that are still supported) should be identical to previous implementations.

Reply:
Different mangling and aliases is part of the solution but that's the easy bit.  The unification rules and how this interacts with structural typing are the difficult bits.